tests/phpunit/integration/Modules/Analytics_4Test.php

```diff
@@ -206,13 +212,672 @@ public function test_get_datapoints() {
+
+		return $analytics->get_data( 'report', $report_params );
+	}
+
+	/**
+	 * Sets up a mock Analytics 4 instance in preparation for retrieving a report.
+	 *
+	 * This is a helper method to avoid duplicating the same code in multiple tests.
+	 * It also allows us to mock the HTTP client to verify the request parameters.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $property_id The property ID.
+	 * @param array $request_handler_calls Output variable for tracking the request handler calls. Passed by reference.
+	 * @param boolean [enable_validation] Whether to enable validation of the metrics and dimensions. Default false.
+	 * @return Analytics_4 The Analytics 4 instance.
+	 */
+	protected function setup_report( $property_id, array &$request_handler_calls, $enable_validation = false ) {
+		$user_id        = $this->factory()->user->create( array( 'role' => 'administrator' ) );
+		$context        = new Context( GOOGLESITEKIT_PLUGIN_MAIN_FILE );
+		$options        = new Options( $context );
+		$user_options   = new User_Options( $context, $user_id );
+		$authentication = new Authentication( $context, $options, $user_options );
+		$analytics      = new Analytics_4( $context, $options, $user_options, $authentication );
+
+		$authentication->get_oauth_client()->get_client()->setHttpClient(
+			new FakeHttpClient() // Returns 200 by default.
+		);
+
+		// Grant scopes so request doesn't fail.
+		$authentication->get_oauth_client()->set_granted_scopes(
+			$analytics->get_scopes()
+		);
+
+		$http_client = new FakeHttpClient();
+		$http_client->set_request_handler(
+			function ( Request $request ) use ( $property_id, &$request_handler_calls ) {
+				$url    = parse_url( $request->getUrl() );
+				$params = json_decode( (string) $request->getBody(), true );
+
+				$request_handler_calls[] = array(
+					'url'    => $url,
+					'params' => $params,
+				);
+
+				if ( 'analyticsdata.googleapis.com' !== $url['host'] ) {
+					return new Response( 200 );
+				}
+
+				switch ( $url['path'] ) {
+					case "/v1beta/properties/$property_id:batchRunReports":
+						// Return a mock report.
+						return new Response(
+							200,
+							array(),
+							Stream::factory(
+								json_encode(
+									array(
+										'kind'    => 'analyticsData#batchRunReports',
+										'reports' => array(
+											array(
+												'rows' => array(
+													array(
+														'metricValues' => array(
+															array(
+																'value' => 'some-value',
+															),
+														),
+													),
+												),
+											),
+										),
+									)
+								)
+							)
+						);
+
+					default:
+						return new Response( 200 );
+				}
+			}
+		);
+
+		$analytics->get_client()->setHttpClient( $http_client );
+		$analytics->register();
+
+		if ( $enable_validation ) {
+			// Metrics and dimensions are only validated when using shared credentials; this block of code sets up the shared credentials scenario.
```

Comment:
Have extracted to a helper method as mentioned. Also, yeah I do prefer to avoid boolean parameters, should have done so here really. It's gone now anyway!